apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/templates-parser.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * External dependencies
+ */
+/* eslint-disable import/no-extraneous-dependencies */
+import { keyBy, mapValues, get } from 'lodash';
+/* eslint-enable import/no-extraneous-dependencies */
+
+/**
+ * WordPress dependencies
+ */
+/* eslint-disable import/no-extraneous-dependencies */
+import { parse as parseBlocks } from '@wordpress/blocks';
+/* eslint-enable import/no-extraneous-dependencies */
+
+/**
+ * Internal dependencies
+ */
+import replacePlaceholders from './replace-placeholders';
+
+const { templates = [], siteInformation = {} } = window.starterPageTemplatesConfig;
+
+let allTemplatesBlockBySlug = mapValues( keyBy( templates, 'slug' ), ( { content, title } ) => ( {
+	blocks: [],
+	content,
+	count: 0,
+	isEmpty: ! content,
+	isParsing: !! content,
+	title,
+} ) );
+
+// Dispatch a global event once all templates have been parsed.
+setTimeout( () => {
+	window.dispatchEvent(
+		new CustomEvent( 'onTemplatesParse', {
+			detail: {
+				templates: allTemplatesBlockBySlug,
+				isParsing: true,
+			},
+		} )
+	);
+}, 0 );
+
+setTimeout( () => {
+	allTemplatesBlockBySlug = mapValues( allTemplatesBlockBySlug, ( { content, title } ) => {
+		const blocks = content ? parseBlocks( replacePlaceholders( content, siteInformation ) ) : [];
+		return {
+			blocks,
+			count: blocks.length,
+			isParsing: false,
+			isEmpty: ! content,
+			title,
+		};
+	} );
+
+	// Dispatch a global event once all templates have been parsed.
+	window.dispatchEvent(
+		new CustomEvent( 'onTemplatesParse', {
+			detail: {
+				templates: allTemplatesBlockBySlug,
+				isParsing: false,
+			},
+		} )
+	);
+}, 50 );
```

Comment:
So we're still effectively parsing in serial? So the main benefit is that we render something to the page immediately and then once the templates are done parsing we "send" that data to the component to re-render with the actual previews.